<a href="https://colab.research.google.com/github/JaeDoo1034/Kaggle-Study/blob/master/Keras_tuner1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install git+https://github.com/keras-team/keras-tuner.git -q

MoA: Keras + KerasTuner best practices¶<br>
This notebook will teach you how to:<br>

1. Use a Keras neural network for the MoA competition
2. Use KerasTuner to find high-performing model configurations
3. Ensemble a few of the top models to generate final predictions

In [1]:
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

In [2]:
print('TF version:', tf.__version__)
print('GPU devices:', tf.config.list_physical_devices('GPU'))

TF version: 2.3.0
GPU devices: [PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]


In this competition, we're looking at 3 CSV files: one for training features, one for training targets (with the same number of entries and a 1:1 match between entries in the features file and those in the targets file), and one for test features. The goal is to predict the targets that correspond to the test features.

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [6]:
train_features_df = pd.read_csv('/content/drive/My Drive/Data/train_features.csv')
train_targets_df = pd.read_csv('/content/drive/My Drive/Data/train_targets_scored.csv')
test_features_df = pd.read_csv('/content/drive/My Drive/Data/test_features.csv')

In [7]:
print('train_features_df.shape:', train_features_df.shape)
print('train_targets_df.shape:', train_targets_df.shape)
print('test_features_df.shape:', test_features_df.shape)

train_features_df.shape: (23814, 876)
train_targets_df.shape: (23814, 207)
test_features_df.shape: (3982, 876)


In [8]:
train_features_df.sample(5)

sig_id cp_type  cp_time cp_dose  ...    c-96    c-97    c-98    c-99
23743  id_ff39c765a  trt_cp       72      D2  ... -0.5465 -2.2300 -4.7440 -1.1350
20813  id_df1a9501d  trt_cp       72      D2  ...  0.1942 -0.1082  0.2801  1.1350
7678   id_526755774  trt_cp       48      D1  ...  0.2668 -1.0670  0.1236 -1.3740
11050  id_76de9b673  trt_cp       48      D1  ...  0.9486  0.0750 -2.1980  0.0962
3869   id_2964a5d4e  trt_cp       24      D1  ... -0.0964 -0.3451  0.4841 -0.0660

[5 rows x 876 columns]

Ok, so we have 2 categorical features (cp_type and cp_dose, which are strings), and everything else is numerical (assuming g-0 to g-99 are homogeneous in type).

We'll use the StringLookup and CategoryEncoding layers to encode the categorical features, and the Normalization layer to normalize the values of the numerical features.

Let's look at the targets:

In [9]:
train_targets_df.sample(5)

sig_id  ...  wnt_inhibitor
15194  id_a33eaa8e9  ...              0
2563   id_1b42f3587  ...              0
17287  id_ba162cfcb  ...              0
5833   id_3eb6c985b  ...              0
3242   id_22c588033  ...              0

[5 rows x 207 columns]

The targets are binary indicators (0 or 1) across 206 different categories. So our model should output a probability score between 0 and 1 (sigmoid activation) across 206 outputs.

The sample submission format matches these expectations:

In [11]:
sample_submission_df = pd.read_csv('/content/drive/My Drive/Data/sample_submission.csv')
sample_submission_df.sample(5)

sig_id  ...  wnt_inhibitor
969   id_3fb843bae  ...            0.5
584   id_2773f584f  ...            0.5
637   id_2a890edee  ...            0.5
2594  id_a5d6565f4  ...            0.5
592   id_27fae2c05  ...            0.5

[5 rows x 207 columns]

Out of 23,814 samples, how often is each of the 206 target indicators positive?

In [12]:
for target_name in list(train_targets_df)[1:]:
  rate = float(sum(train_targets_df[target_name])) / len(train_targets_df)
  print('%.4f percent positivity rate for %s' % (100*rate, target_name) )

0.0714 percent positivity rate for 5-alpha_reductase_inhibitor
0.0756 percent positivity rate for 11-beta-hsd1_inhibitor
0.1008 percent positivity rate for acat_inhibitor
0.7979 percent positivity rate for acetylcholine_receptor_agonist
1.2640 percent positivity rate for acetylcholine_receptor_antagonist
0.3065 percent positivity rate for acetylcholinesterase_inhibitor
0.2268 percent positivity rate for adenosine_receptor_agonist
0.4031 percent positivity rate for adenosine_receptor_antagonist
0.0504 percent positivity rate for adenylyl_cyclase_activator
1.1338 percent positivity rate for adrenergic_receptor_agonist
1.5117 percent positivity rate for adrenergic_receptor_antagonist
0.2771 percent positivity rate for akt_inhibitor
0.0294 percent positivity rate for aldehyde_dehydrogenase_inhibitor
0.1764 percent positivity rate for alk_inhibitor
0.0504 percent positivity rate for ampk_activator
0.0504 percent positivity rate for analgesic
0.2016 percent positivity rate for androgen_recep

Two things:

- Positivity rates are very low
- Positivity rates are very heterogeneous

Setting aside a validation set
Let's set aside a training set and a validation set: all of our configuration choices will be guided by performance on this subset of the total available training data. We will also keep on the total available training data, which we will use to train our final production models.

In [13]:
num_train_samples = int(0.8 * len(train_features_df))

full_train_features_ids = train_features_df.pop('sig_id')
full_test_features_ids = test_features_df.pop('sig_id')
train_targets_df.pop('sig_id')

full_train_features_df = train_features_df.copy()
full_train_targets_df = train_targets_df.copy()

val_features_df = train_features_df[num_train_samples:]
train_features_df = train_features_df[:num_train_samples]
val_targets_df = train_targets_df[num_train_samples:]
train_targets_df = train_targets_df[:num_train_samples]

print('Total training samples:', len(full_train_features_df))
print('Training split samples:', len(train_features_df))
print('Validation split samples:', len(val_features_df))

Total training samples: 23814
Training split samples: 19051
Validation split samples: 4763


<b>A dumb baseline</b><br>
If you've read my book, you know you should start tough projects by computing a "dumb" baseline that will serve as your reference point. This is usually the highest score you can reach without looking at the test features (or validation features in this case). Let's use the positivity rate of each target as measured in the training subset to generate predictions for the validation subset.

In [14]:
predictions = []
for target_name in list(train_targets_df):
  rate = float(sum(train_targets_df[target_name])) / len(train_targets_df)
  predictions.append(rate)

predictions = np.array([predictions] * len(val_features_df))

targets = np.array(val_targets_df)
score = keras.losses.BinaryCrossentropy()(targets,predictions)
print('Baseline score : %.4f' % score.numpy())

Baseline score : 0.0209


<b>Prepare TF datasets</b><br>
Let's turn our dataframes into tf.data.Datasets, which we will use to train our Keras models in the next step. Our datasets will yield tuples of (features, targets) where features is a dict and targets is a list. In the features dict, we will have 3 keys: cp_type and cp_dose, as well as numerical_features, which will be a vector concatenating all numerical features in the space.

In [18]:
feature_names = list(train_features_df)
categorical_feature_names = ['cp_type', 'cp_dose']
numerical_feature_names = [name for name in feature_names if name not in categorical_feature_names]

def merge_numerical_features(feature_dict):
    categorical_features = {name: feature_dict[name] for name in categorical_feature_names}
    numerical_features = tf.stack([tf.cast(feature_dict[name], 'float32') for name in numerical_feature_names])
    feature_dict = categorical_features
    feature_dict.update({'numerical_features': numerical_features})
    return feature_dict

tf.data : TensorFlow 입력 파이프 라인 빌드<br>
https://www.tensorflow.org/guide/data?hl=ko

In [21]:
train_features_ds = tf.data.Dataset.from_tensor_slices(dict(train_features_df))
train_features_ds = train_features_ds.map(lambda x: merge_numerical_features(x))

train_targets_ds = tf.data.Dataset.from_tensor_slices(np.array(train_targets_df))
train_ds = tf.data.Dataset.zip((train_features_ds, train_targets_ds))

full_train_features_ds = tf.data.Dataset.from_tensor_slices(dict(full_train_features_df))
full_train_features_ds = full_train_features_ds.map(lambda x: merge_numerical_features(x))


In [35]:
# dict형태로 준비하네..
train_features_ds

<MapDataset shapes: {cp_type: (), cp_dose: (), numerical_features: (873,)}, types: {cp_type: tf.string, cp_dose: tf.string, numerical_features: tf.float32}>

In [29]:
full_train_features_ds

<MapDataset shapes: {cp_type: (), cp_dose: (), numerical_features: (873,)}, types: {cp_type: tf.string, cp_dose: tf.string, numerical_features: tf.float32}>

In [34]:
val_features_ds = tf.data.Dataset.from_tensor_slices(dict(val_features_df))
# dict형태로 데이터를 준비하네 신기..
val_features_ds =val_features_ds.map(lambda x: merge_numerical_features(x))

val_targets_ds = tf.data.Dataset.from_tensor_slices(np.array(val_targets_df))
val_ds = tf.data.Dataset.zip((val_features_ds,val_targets_ds))

test_ds = tf.data.Dataset.from_tensor_slices(dict(test_features_df))
test_ds = test_ds.map(lambda x: merge_numerical_features(x))

In [39]:
# cardinalty() 참고자료 : https://www.tensorflow.org/api_docs/python/tf/data/Dataset
# 집합의 크기를 표시
print('Training split samples :', int(train_ds.cardinality()))
print('Validation split samples:',int(val_ds.cardinality()))
print('Test samples:',int(test_ds.cardinality()))

Training split samples : 19051
Validation split samples: 4763
Test samples: 3982


In [42]:
train_ds = train_ds.shuffle(1024).batch(64).prefetch(8)

In [43]:
train_ds

<PrefetchDataset shapes: ({cp_type: (None,), cp_dose: (None,), numerical_features: (None, 873)}, (None, 206)), types: ({cp_type: tf.string, cp_dose: tf.string, numerical_features: tf.float32}, tf.int64)>

20년 9월 30일 작업 완료 : 텐서플로우 파이프라인 구축